Qt6

samaaron opened the pull request:
First pass at getting Sonic Pi to build on Qt6.

Appears to work on Windows - needs testing on other platforms.

samaaron committed:
GUI - Qt6 - update CMakeLists files to work with Qt6

They *should* fall back to Qt5 if Qt6 isn't available. This likely requires a very recent Qt5 to work correctly such as Qt 5.15 and may need some tweaking to work correctly on all platforms

samaaron committed:
GUI - Qt6 - switch to using QRecursiveMutex

A QMutex in "recursive mode" doesn't appear to exist in Qt6 - however QRecursiveMutex is in both Qt6 and Qt5.15

samaaron committed:
GUI - Qt6 - use more recent size policy

samaaron committed:
GUI - Qt6 - update how encoding is set to UTF-8 for Qt6 compatibility

samaaron committed:
GUI - Qt6 - update qss theme to fix text visibility issue within Qt6

This needs testing across all platforms including on Qt5 to ensure it doesn't introduce other issues

samaaron committed:
GUI - Qt6  - ensure scope.h can be found

samaaron committed:
GUI - Qt6 - fix helpPages keyword test by using empty string rather than NULL for default value

samaaron committed:
GUI - Qt6 - re-implement full screen mode

On Windows when in full-screen mode the menus and drop downs aren't visible due to limitations with OpenGL rendering. This could be fixed in Qt5 by setting QWindowsWindowFunctions::setHasBorderInFullScreen. However, this doesn't appear to be available in Qt 6.

Instead we simply drop the window chrome and set the size of the window to match the current monitor (whilst remembering the previous size). When we switch back out of full-screen mode we revert back to the previous size and position.

This needs to be tested on macOS and Linux

samaaron committed:
GUI - Qt6 - fix issue with Tab key under Qt6 on Windows

Built with Qt6, the tab key doesn't work.

This is fixed by moving from using a signalMapper (which I was likely using incorrectly) to a nice modern C++ lambda and more modern QKeySequence description for Tab.

samaaron committed:
GUI - Qt6 - it appears to not be correct in Qt6 to set margins to 0

(I'm not entirely sure what this was supposed to achieve in the first place!)

samaaron committed:
GUI - Qt6 - absolutePath appears to be part of QDir but no longer QFileInfo. Adjust accordingly

samaaron committed:
GUI - Qt6 - fix compiler warning about for loop counter type

samaaron committed:
Docs - update Windows build instructions for Qt6

Also update for more recent versions of Elixir and Ruby

samaaron committed:
Github - use Qt6 for CI

Should still support Qt5 on Linux, so need to test on both for that platform.

ethancrawford commented on BUILD-WINDOWS.md:
```diff
@@ -96,24 +96,19 @@ Ruby is needed both for a number of the build steps and as the main
-processes, rugged which is used to store the code diffs in a local Git
-repo and ffi which is used to hook into external shared libs such as
-libaubio.
+This uses the Ruby library management tool `gem` rugged which is used to
```
This edit seems a bit mixed up to me - is it correct?

samaaron replied:
Good catch, I've fixed it now :-)

samaaron committed:
Github - be more specific with Qt version

samaaron committed:
Build - fix description of rugged

lilyinstarlight commented on app/gui/qt/CMakeLists.txt:
```diff
@@ -30,13 +30,20 @@ endif()
 
-find_package(Qt5 COMPONENTS Core Widgets Gui Concurrent Network OpenGL PrintSupport Xml Svg WebEngineWidgets REQUIRED)
+
+find_package(Qt6 COMPONENTS Widgets Core Widgets Gui Concurrent Network OpenGLWidgets PrintSupport Xml Svg WebEngineWidgets REQUIRED)
```
It looks like `Widgets` got specified twice. That's not necessary, right?

```suggestion
find_package(Qt6 COMPONENTS Core Widgets Gui Concurrent Network OpenGLWidgets PrintSupport Xml Svg WebEngineWidgets REQUIRED)
```

samaaron replied:
haha, no!

samaaron committed:
Github - Actions - update Qt cache name and path

Wondering whether this will cause a refresh and help the build. Unlikely, but worth a shot!

samaaron committed:
Build - remove duplicate Widgets declaration

samaaron committed:
Github - CI - use more recent Ruby on Windows and no longer install unnecessary gems

SunderB commented:
I've not tried it myself, but looking at the ubuntu CI builds, it seems Qt6WebEngineWidgets is missing.
On Debian at least it seems like Qt6 web engine is still being packaged, and isn't in the debian repo yet: https://salsa.debian.org/qt-kde-team/qt6/qt6-webengine.

samaaron committed:
Github - try changing qt module name to see if it fixes CI error

samaaron committed:
Build - make webengine an optional feature disabled by default

The webengine feature now requires Qt6 and a the following cmake flag set: -D WITH_QT_GUI_WEBENGINE=ON

samaaron committed:
Docs - update build instructions for all platforms

samaaron committed:
Build - remove webengine stuff from Pi setup script

This should wait for Qt6 and proper support on Raspberry Pi OS - hopefully in the not too distant future...

ethancrawford commented:
@samaaron there were typos in the last few commits 😛

I will make a PR soon...
